nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -243,6 +245,113 @@ impl Builtin for StringPrototypeIterator {
+/// and returns either a normal completion containing a String or a throw
+/// completion.
+fn string_padding_builtins_impl(
+    agent: &mut Agent,
+    o: Value,
+    max_length: Value,
+    fill_string: Value,
+    placement_start: bool,
+) -> JsResult<Value> {
+    // 1. Let S be ? ToString(O).
+    let s = to_string(agent, o)?;
+
+    // 2. Let intMaxLength be ℝ(? ToLength(maxLength)).
+    let int_max_length = to_length(agent, max_length)?;
+
+    // 3. Let stringLength be the length of S.
+    let string_length = s.utf16_len(agent) as i64;
+
+    // 4. If intMaxLength ≤ stringLength, return S.
+    if int_max_length <= string_length {
+        return Ok(s.into());
+    }
+
+    // 5. If fillString is undefined, set fillString to the String value consisting solely of the code unit 0x0020 (SPACE).
+    let fill_string = if fill_string.is_undefined() {
+        BUILTIN_STRING_MEMORY.r#__
+    } else {
+        // 6. Else, set fillString to ? ToString(fillString).
+        to_string(agent, fill_string)?
+    };
+
+    // 7. Return StringPad(S, intMaxLength, fillString, placement).
+    string_pad(agent, s, int_max_length, fill_string, placement_start)
+}
+
+/// ### [22.1.3.17.2 StringPad ( S, maxLength, fillString, placement )](https://tc39.es/ecma262/#sec-stringpad)
+///
+/// The abstract operation StringPad takes arguments S (a String),
+/// maxLength (a non-negative integer), fillString (a String), and
+/// placement (start or end) and returns a String.
+fn string_pad(
+    agent: &mut Agent,
+    s: String,
+    max_len: i64,
+    fill_string: String,
+    placement_start: bool,
+) -> JsResult<Value> {
+    // 1. Let stringLength be the length of S.
+    let string_length = s.utf16_len(agent) as i64;
+
+    // 2. If maxLength ≤ stringLength, return S.
+    if max_len <= string_length {
+        return Ok(s.into());
+    }
+
+    // 3. If fillString is the empty String, return S.
+    if fill_string.is_empty_string() {
+        return Ok(s.into());
+    }
+
+    // 4. Let fillLen be maxLength - stringLength.
+    let fill_len = max_len - string_length;
+    let fill_string_length = fill_string.utf16_len(agent) as i64;
+
+    // 5. Let truncatedStringFiller be the String value consisting of repeated concatenations of fillString truncated to length fillLen.
+    let mut strings = if fill_len == fill_string_length {
+        let mut vec = VecDeque::with_capacity(2);
+        vec.push_back(fill_string);
+        vec
+    } else if fill_len % fill_string_length == 0 {
+        let fill_count = (fill_len / fill_string_length) as usize;
+        let mut vec = VecDeque::with_capacity(fill_count + 1);
+        vec.extend(repeat(fill_string).take(fill_count));
+        vec
+    } else if fill_len < fill_string_length {
+        let mut vec = VecDeque::with_capacity(2);
+        // TODO: Deal with surrogates.
+        let sub_string = &fill_string.as_str(agent)[..fill_len as usize];
```

Comment:
**thought:** This is actually probably not quite correct. The length values are UTF-16 indexes but then we're indexing into a UTF-8 string with those. These should probably go through the index mapping.

Reply:
Pushed what I think is a fix for this, doesn't really change any expectations but now the crashes point to https://github.com/trynova/nova/blob/main/small_string/lib.rs#L40 which is more correct.

Pretty messy code which might be cleaned up by some iterator magic but I couldn't figure it out.